designs/2024-hooks-for-test-cases/README.md

```diff
@@ -0,0 +1,311 @@
+    burden, incompatibility with other tools, breaking existing user
+    experience, etc. Try to identify as many potential problems with
+    implementing this RFC as possible.
+-->
+
+Possible concerns:
+
+- Potential for Misuse:
+  - Despite being an optional feature, there remains a risk that some users might misuse the setup property
+    by inadvertently modifying global state, leading to flaky tests.
+- Performance:
+  - There is a risk that developers might overload the setup property, leading to performance decrease.
+- Redundancy: confusion with `beforeEach`:
+  - It is important to emphasize the key difference between these entities in the documentation:
+    - `beforeEach` is for **consistent** actions applicable to **all** tests;
+    - `setup` is for **specific** actions applicable to **particular** test.
+- Limited Use Case:
+  - The necessity for the `setup` might be limited to specific rules and plugins. The broader ESLint user
+    community may not frequently encounter these challenges, and it may seem unjustified to them.
+
+## Backwards Compatibility Analysis
+
+<!--
+    How does this change affect existing ESLint users? Will any behavior
+    change for them? If so, how are you going to minimize the disruption
+    to existing users?
+-->
+
+Adding new optional property to the test case is backwards compatible.
+When the property is not set, `RuleTester` acts the same for existing users — no disruption expected.
+
+## Alternatives
+
+<!--
+    What other designs did you consider? Why did you decide against those?
+
+    This section should also include prior art, such as whether similar
+    projects have already implemented a similar feature.
+-->
+
+My current workaround is assigning a custom `it` function to the `RuleTester` that looks up for a mock by test name:
+
+```javascript
+const readerMock = jest.fn();
+jest.mock('fs', () => ({
+  readFileSync: readerMock
+}));
+
+const mockedEnvironments = {
+  "test case one": { dependencies: { "some-package": "^1.0.0" } },
+  "test case two": { devDependencies: { "another-package": "^2.0.0" } }
+}
+
+RuleTester.it = (name, ...rest) => {
+  readerMock.mockReturnValue(
+    JSON.stringify( mockedEnvironments[name] )
+  );
+  it(name, ...rest);
+};
+
+new RuleTester().run("my-custom-rule", myCustomRule, {
+  valid: [
+    {
+      name: "test case one",
+      code: '/* valid code example */',
+      options: [/* some options */]
+    }
+  ],
+  invalid: [
+    {
+      code: '/* invalid code example */',
+      errors: [{ messageId: "someErrorId" }],
+      options: [/* some options */]
+    }
+  ]
+});
+```
+
+However, I'm not satisfied with this approach due to a number of disadvantages:
+
+- The `name` property is required to be set on each test case so that `it` could distinguish between them;
+- Also, that `name` must be unique for proper lookup;
+- And most importantly: mocked environment is detached from the test cases definitions:
```

Comment:
A workaround where the setup can be specified along with the tests can be to directly use features of the testing framework through which you run the tests. For example, with `mocha` or `jest`, you could use `beforeEach` and `afterEach`: 

```js
const ruleTester = new RuleTester();

describe("My custom rule with some setup", () => {
    beforeEach(() => {
        // some setup
    });

    afterEach(() => {
        // cleanup
    });

    ruleTester.run("my-custom-rule", myCustomRule, {
        valid: [
            {
                code: "/* valid code example */",
                options: [
                    /* some options */
                ],
            },
        ],
        invalid: [
            {
                code: "/* invalid code example */",
                errors: [{ messageId: "someErrorId" }],
                options: [
                    /* some options */
                ],
            },
        ],
    });
});

describe("My custom rule with another setup", () => {
    beforeEach(() => {
        // another setup
    });

    afterEach(() => {
        // cleanup
    });

    ruleTester.run("my-custom-rule", myCustomRule, {
        valid: [
            {
                code: "/* another valid code example */",
                options: [
                    /* some options */
                ],
            },
        ],
        invalid: [
            {
                code: "/* another invalid code example */",
                errors: [{ messageId: "someErrorId" }],
                options: [
                    /* some options */
                ],
            },
        ],
    });
});

```

Reply:
`beforeEach` is before EACH — not a unique setup before a particular case.

Reply:
Yes, but you can split test runs into multiple `describe`-s where each has its own `beforeEach` so each test could have its own setup, as in the above example. It's a common practice with testing frameworks.

Reply:
One can live with a single hand, but why would they do that if they could have both?

Reply:
also, both `valid` AND `invalid` are required, so it can not be a single case in every `describe`, @mdjermanovic